src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,1079 @@
+                {
+                    aNumPreferredEntries++;
+                }
+            }
+
+            if (serverSubTlv->GetServerDataLength() >= sizeof(Service::DnsSrpUnicast::ServerData))
+            {
+                aNumEntries++;
+
+                // If our entry also uses the server TLV data (with
+                // ML-EID address), then the we prefer based on the
+                // associated RLOC16.
+
+                if ((GetType() == kTypeUnicastMeshLocalEid) && IsPreferred(serverSubTlv->GetServer16()))
+                {
+                    aNumPreferredEntries++;
+                }
+            }
+        }
+    }
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// Publisher::DnsSrpServiceEntry::Info
+
+Publisher::DnsSrpServiceEntry::Info::Info(Type aType, uint16_t aPortOrSeqNumber, const Ip6::Address *aAddress)
+    : mPortOrSeqNumber(aPortOrSeqNumber)
+    , mType(aType)
+{
+    // It is important to `Clear()` the object since we compare all
+    // bytes using overload of operator `==`.
+
+    Clear();
+
+    mType            = aType;
+    mPortOrSeqNumber = aPortOrSeqNumber;
+
+    if (aAddress != nullptr)
+    {
+        mAddress = *aAddress;
+    }
+}
+
+#endif // OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
+
+//---------------------------------------------------------------------------------------------------------------------
+// Publisher::PrefixEntry
+
+void Publisher::PrefixEntry::Publish(const OnMeshPrefixConfig &aConfig)
+{
+    otLogInfoNetData("Publisher: Publishing OnMeshPrefix %s", aConfig.GetPrefix().ToString().AsCString());
+
+    mType   = kTypeOnMeshPrefix;
+    mPrefix = aConfig.GetPrefix();
+    mFlags  = aConfig.ConvertToTlvFlags();
+
+    SetState(kToAdd);
+
+    Process();
+}
+
+void Publisher::PrefixEntry::Publish(const ExternalRouteConfig &aConfig)
+{
+    otLogInfoNetData("Publisher: Publishing ExternalRoute %s", aConfig.GetPrefix().ToString().AsCString());
+
+    mType   = kTypeExternalRoute;
+    mPrefix = aConfig.GetPrefix();
+    mFlags  = aConfig.ConvertToTlvFlags();
+
+    SetState(kToAdd);
+
+    Process();
+}
+
+void Publisher::PrefixEntry::Unpublish(void)
+{
+    bool registerWithLeader = false;
+
+    otLogInfoNetData("Publisher: Unpublishing %s", mPrefix.ToString().AsCString());
+
+    registerWithLeader = Remove(/* aNextState */ kNoEntry);
+
+    if (registerWithLeader)
+    {
+        Get<Notifier>().HandleServerDataUpdated();
+    }
+}
+
+void Publisher::PrefixEntry::HandleNotifierEvents(Events aEvents)
+{
+    if (aEvents.ContainsAny(kEventThreadNetdataChanged | kEventThreadRoleChanged))
+    {
+        Process();
+    }
+}
+
+bool Publisher::PrefixEntry::Add(void)
+{
+    // Adds the prefix entry to the network data.
+
+    bool registerWithLeader = false;
+
+    union
+    {
+        OnMeshPrefixConfig  mOnMeshPrefix;
+        ExternalRouteConfig mExternalRoute;
+    } config;
```

Comment:
We use `union` here to reduce the use of stack in this method (the two configs are large objects and we only need one). 

This matters because from `Add()` we can call user's callback to notify that entry was added (`mPrefixCallback` from `NotifyPrefixEntryChange()`). Whenever we invoke a user callback it's good to optimize our stack use if we can (leave more stack for user to use).

Reply:
aha that makes sense to me!
Maybe leave a comment to describe the intention?

Reply:
I actually changed this in new push, added two new small methods `AddOnMeshPrefix()` and `AddExternalRoute()` which we then call from `Add()`. This helps moves the `config` definitions to the new methods and their stack use should be freed before we invoke user callback.